Work around freeze due to VP9 hardware decode, #4486

low-batt opened the pull request:
This commit will add code to MPVController.mpvInit that changes the value of the mpv option hwdec-codecs, removing VP9, if IINA is not running on a Mac with an Apple Silicon chip. This change works around the FFmpeg defect reported in ticket 9599.

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [x] This implements/fixes issue #4486.

---

**Description:**

low-batt committed:
Workaround freeze due to VP9 hardware decode, #4486

This commit will add code to MPVController.mpvInit that changes the
value of the mpv option hwdec-codecs, removing VP9, if IINA is not
running on a Mac with an Apple Silicon chip. This change works around
the FFmpeg defect reported in ticket 9599.

low-batt commented:
I was at the eye doctor today, so eyes dilated and things are blurry. A critical review is needed. Especially the detection of Apple Silicon. Is there a better way to do that?  So far we have not seen the problem under Apple Silicon. so the workaround is not applied when running on AS and VP9 hardware decoding is still used. Does this make sense, or should we always disable VP9 for now to be safe?

svobs commented:
> Especially the detection of Apple Silicon. Is there a better way to do that?

Not really. Seems like there should be, but did some searching and yours appears to be the standard solution. Very typical to see Linux shell scripts rely on parsing the `uname` command, so this isn't unexpected.

Have not tried running it yet, but looks like a good solution. +1 for adding a secret pref allowing the workaround to be disabled.

Also, you should know the word "workaround" (with no space) is only used as a noun. It doesn't have an exact equivalent verb in common English, although you could use "work around".

uiryuu commented on iina/MPVController.swift:
```diff
@@ -153,6 +153,66 @@ class MPVController: NSObject {
+      Logger.log("Failed to construct string for sysinfo.machine", level: .error)
+      return false
+    }
+    return machine.replacingOccurrences(of: "\0", with: "") == "arm64"
```
I suggest use `machine.starts(with: "arm64")`, which is not only more robust but also more elegant.

low-batt replied:
Good catch. Uploaded a new commit with this change.

uiryuu reviewed (changes requested):
Other changes LGTM

low-batt committed:
Work around freeze due to VP9 hardware decode, #4486

This commit will add code to MPVController.mpvInit that changes the
value of the mpv option hwdec-codecs, removing VP9, if IINA is not
running on a Mac with an Apple Silicon chip. This change works around
the FFmpeg defect reported in ticket 9599.

low-batt commented:
On this:
> +1 for adding a secret pref allowing the workaround to be disabled.

I thought about that, but instead I made sure that if this mpv option is set in `Advanced` settings, the workaround will not override the value configured there. In fact the workaround detects that a value has been configured for the option and does not apply the workaround. I did that so that it was clear the workaround is not being used when you read the log file.

I was wondering if we should have an internal preference that by passes the check for Apple Silicon and always applies the workaround. Just in case we find it also has trouble on Apple Silicon machines. I did not do that as so far I have not been able trigger a problem on my M1 machine.
